Allow video to be recorded till the end of the test

bahmutov opened the pull request:
Adds hardcoded delay of 1 second after the end of the test run before closing the browser if there is video recording going on. This allows the last frames to be added, creating the complete video.

- Closes #2370
- added separate debug stream `DEBUG=cypress:server:video:frames` for Electron video frames, which should give more insight into video freezing in #4722  
- [x] fully records video in https://github.com/bahmutov/cypress-test-clock
- [x] fully records the repo from the original issue https://github.com/PinkyJie/cypress-auto-stub-example

## Reviews

Trying something new

- [x] initial review by @Bkucera 
- [x] final review by @brian-mann 

## Before

the last frame of the video

<img width="1277" alt="Screen Shot 2019-07-24 at 10 54 58 AM" src="https://user-images.githubusercontent.com/2212006/61804341-c2540f80-ae01-11e9-86fa-9c7a53db69d4.png">

## After

<img width="1280" alt="Screen Shot 2019-07-24 at 10 56 01 AM" src="https://user-images.githubusercontent.com/2212006/61804364-cc760e00-ae01-11e9-866c-c9384297ece4.png">

### Pre-merge Tasks

<!-- The following tasks must be completed before a PR can be merged.
You can delete tasks if they are not applicable to the PR changes. -->

- [x] Have tests been added/updated for the changes in this PR?
- [x] Has the original issue been tagged with a release in ZenHub? <!-- (internal team only)-->

bahmutov committed:
WIP: record the video fully without cutting the end of the test

bahmutov committed:
delay closing browser when recording video to avoid chopping off the end

bahmutov committed:
refactor code and update tests

bahmutov committed:
do not wrap delay method twice

Bkucera commented on packages/server/lib/modes/run.coffee:
```diff
@@ -677,10 +680,28 @@ module.exports = {
+      # and there still might be frames sent by the browser
+      # in order to avoid chopping off the end of the video
+      # delay closing the browser by N ms
+      DELAY_TO_LET_VIDEO_FINISH_MS = 1000
```
you could instead export + stub this delay value in the test to get extra coverage of this function

Bkucera commented on packages/server/test/unit/modes/run_spec.coffee:
```diff
@@ -288,6 +288,7 @@ describe "lib/modes/run", ->
   context ".waitForTestsToFinishRunning", ->
     beforeEach ->
       sinon.stub(@projectInstance, "getConfig").resolves({})
+      sinon.stub(runMode, "_delayToLetVideoFinish").resolves()
```
why not stub the delay value instead of the function?

Bkucera replied:
well, i guess you would need to `spy` on the function either way

bahmutov replied:
yeah, making it a method allows spying and stubbing on it during tests, rather than exposing a constant - plus the logic for waiting could be dynamic in the future (like waiting for a browser event for example)

jennifer-shehane commented:
eek. I know this likely solves the issue, but...hardcoded 1 sec wait is kind of the opposite of Cypress philosophy. This will slow down user's runs, especially those that have 100+ spec files. Would rather isolate the cause and somehow hook into knowing when the reporter is *done* rendering to stop the video there.

bahmutov commented:
yeah, imperfect, but good enough for now to fix the immediate problem. The video recording will probably change how we capture and end the video. 1-second addition per spec is a fair price I feel.

bahmutov committed:
add extra verbose frames logging

brian-mann reviewed (changes requested):
We should only delay the end of the video if we are actually recording a video

bahmutov commented:
@brian-mann in the file it checks if we are recording the video, if there is a better way
```

    .tap =>
      # ? is this the best way to determine if we are recording a video?
      # TODO rename "end" to something meaningful, like "videoCaptureEnd"
      return unless end
```

brian-mann commented:
Def agree with your comments on making this much clearer

brian-mann committed:
rename video capture props to be more discernible, refactor code, general cleanup

- simplify promise implementation in videoCapture.start
- extract a few methods in runMode to handle conditional video capturing

brian-mann committed:
fix failing test

brian-mann committed:
cleanup debug logs
